config/management/src/config_builder.rs

```diff
@@ -0,0 +1,270 @@
+        let layout_value = Value::String(layout.to_toml().unwrap());
+        common_storage
+            .set(crate::constants::LAYOUT, layout_value)
+            .unwrap();
+    }
+
+    /// Association initializes its account and key.
+    fn create_association(&self) {
+        self.storage_helper.initialize(ASSOCIATION.into());
+        self.storage_helper
+            .association_key(ASSOCIATION, ASSOCIATION_SHARED)
+            .unwrap();
+    }
+
+    /// Operators upload their keys to shared storage
+    /// Operators initialize their validator endpoint and full node endpoint
+    /// Operators upload their validator_config to shared storage
+    fn initialize_validator_config(&self, index: usize) -> NodeConfig {
+        let ns = index.to_string();
+        let ns_shared = ns.clone() + SHARED;
+        self.storage_helper.initialize(ns.clone());
+
+        let operator_key = self.storage_helper.operator_key(&ns, &ns_shared).unwrap();
+
+        let validator_account = account_address::from_public_key(&operator_key);
+        let mut config = self.template.clone_for_template();
+        config.randomize_ports();
+
+        let validator_network = config.validator_network.as_mut().unwrap();
+        let validator_network_address = validator_network.listen_address.clone();
+        validator_network.identity = Identity::from_storage(
+            libra_global_constants::VALIDATOR_NETWORK_KEY.into(),
+            libra_global_constants::OPERATOR_ACCOUNT.into(),
+            self.secure_backend(&ns, "validator"),
+        );
+
+        let fullnode_network = &mut config.full_node_networks[0];
+        let fullnode_network_address = fullnode_network.listen_address.clone();
+        fullnode_network.identity = Identity::from_storage(
+            libra_global_constants::FULLNODE_NETWORK_KEY.into(),
+            libra_global_constants::OPERATOR_ACCOUNT.into(),
+            self.secure_backend(&ns, "full_node"),
+        );
+
+        self.storage_helper
+            .validator_config(
+                validator_account,
+                validator_network_address,
+                fullnode_network_address,
+                &ns,
+                &ns_shared,
+            )
+            .unwrap();
+        config
+    }
+
+    /// Operators generate genesis from shared storage and verify against waypoint.
+    /// Insert the genesis/waypoint into local config.
+    fn finish_validator_config(&self, index: usize, config: &mut NodeConfig) {
+        let ns = index.to_string();
+        let genesis_path = TempPath::new();
+        genesis_path.create_as_file().unwrap();
+        let genesis = self.storage_helper.genesis(genesis_path.path()).unwrap();
+        self.storage_helper
+            .insert_waypoint(&ns, constants::COMMON_NS)
+            .unwrap();
+        let output = self
+            .storage_helper
+            .verify_genesis(&ns, genesis_path.path())
+            .unwrap();
+        assert_eq!(output.split("match").count(), 5);
```

Comment:
we have 4 outputs, so when we split we should expect 5 values, e.g., splitting "hello,you" at "," would get "hello" "you"

Reply:
the magic 5 is the same as num_validators you use and I was confused